File: leetcode/001_two_sum/leetcode_0001_20180807.java
import java.util.Map;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] ans = new int[2];
        Map<Integer, List<Integer> > myMap = new HashMap<>();
        for(int i = 0;i < nums.length;i++){
            if(!myMap.containsKey(nums[i])){
                List<Integer> list = new ArrayList<Integer>();
                list.add(i);
                myMap.put(nums[i], list);
            }else{
                myMap.get(nums[i]).add(i);
            }
        }
        for(Map.Entry<Integer, List<Integer>> entry: myMap.entrySet()){
            int temp = target - entry.getKey();
            List<Integer> otherList = myMap.get(temp);
            if(otherList != null){
                ans[0] = entry.getValue().get(0);
                if(otherList == entry.getValue()){
                    ans[1] = entry.getValue().get(1);
                }else {
                    ans[1] = otherList.get(0);
                }
                break;
            }
        }
        if(ans[0] > ans[1]){
            int tmp = ans[0];
            ans[0] = ans[1];
            ans[1] = tmp;
        }
        return ans;
    }

    public static void main(String[] args){
        Solution so = new Solution();
        int nums[] = {3,2,4};
        int result[] = so.twoSum(nums, 6);
        System.out.println(result[0]);
        System.out.println(result[1]);
    }
}

File: leetcode/001_two_sum/leetcode_0001_20180807.cpp
#include <iostream>
#include <vector>
#include <map>
#include <iterator>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target){
        map<int, vector<int> > myMap;
        for (int i = 0; i < nums.size();i++){
            if(myMap.find(nums[i]) == myMap.end()){
                vector<int> ve;
                ve.push_back(i);
                myMap[nums[i]] = ve;
            } else {
                myMap[nums[i]].push_back(i);
            }
        }
        vector<int> ans;
        for (map<int, vector<int> >::iterator it = myMap.begin(); it != myMap.end();it++){
            int temp = target - it->first;
            map<int, vector<int> >::iterator it1 = myMap.find(temp);
            if(it1 != myMap.end()) {
                ans.push_back(it->second[0]);
                if(it != it1) {
                    ans.push_back(it1->second[0]);
                }else {
                    ans.push_back(it->second[1]);
                }
                break;
            }
        }
        sort(ans.begin(), ans.end());
        return ans;
    }
};

int main() {
    cout << "Hello world!!" << endl;

    Solution so;
    vector<int> ve;
    ve.push_back(2);
    ve.push_back(7);
    ve.push_back(3);
    ve.push_back(15);
    vector<int> ans = so.twoSum(ve, 6);
    for (int i = 0; i < ans.size();i++){
        cout << ans[i] << " ";
    }
    cout << endl;
    cin >> ans[0];
    return 0;
}